File: ics-runtime/ics_runtime/core/agent.py
"""Agent — the top-level developer-facing object for ICS Runtime."""

from __future__ import annotations

from typing import Any, Callable, TYPE_CHECKING

from ics_runtime.contracts.capability_enforcer import CapabilityEnforcer
from ics_runtime.contracts.output_contract import OutputContract
from ics_runtime.core.session import Session, SessionContext
from ics_runtime.prompt.builder import PromptBuilder
from ics_runtime.session_backends.memory import MemoryBackend
from ics_runtime.session_backends.base import SessionBackend
from ics_runtime.tools.registry import ToolRegistry

if TYPE_CHECKING:
    from ics_runtime.providers.base import ProviderBase


def _make_provider(provider: str, model: str, api_key: str | None, **kwargs: Any) -> "ProviderBase":
    if provider == "anthropic":
        from ics_runtime.providers.anthropic import AnthropicProvider
        return AnthropicProvider(model=model, api_key=api_key, **kwargs)
    elif provider == "openai":
        from ics_runtime.providers.openai import OpenAIProvider
        return OpenAIProvider(model=model, api_key=api_key, **kwargs)
    raise ValueError(
        f"Unknown provider '{provider}'. Supported: 'anthropic', 'openai'."
    )


_DEFAULT_MODELS = {
    "anthropic": "claude-3-5-sonnet-20241022",
    "openai": "gpt-4o",
}


class Agent:
    """Top-level developer-facing object.  Holds the static ICS configuration
    and spawns ``Session`` instances for each conversation.

    The Agent itself is stateless — all mutable conversation state lives in
    ``Session``.  Creating an Agent is cheap and can be done at module level.

    Args:
        provider:        ``"anthropic"`` or ``"openai"``.
        system:          Shorthand for ``immutable``.  Convenience for simple
                         agents without separate capability declarations.
        immutable:       Text for the IMMUTABLE_CONTEXT ICS layer.
        capability:      Text for the CAPABILITY_DECLARATION ICS layer.
                         ALLOW/DENY/REQUIRE directives go here.
        model:           Provider model identifier.  Defaults to the provider's
                         recommended model if omitted.
        tools:           List of ``@tool``-decorated callables.
        output_contract: ``OutputContract`` instance for post-execution
                         schema validation.
        session_backend: Storage backend for session state.  Defaults to
                         ``MemoryBackend`` (in-process, no persistence).
        api_key:         API key for the provider.  Falls back to the standard
                         environment variable (``ANTHROPIC_API_KEY`` /
                         ``OPENAI_API_KEY``) if not supplied.

    Example::

        agent = Agent(
            provider="anthropic",
            immutable="You are a BFSI lead qualification assistant.",
            capability="DENY: logging PII\\nREQUIRE: risk category on every result",
            tools=[lookup_lead],
            output_contract=OutputContract(schema=QualificationResult),
        )

        with agent.session(lead_id="L-42") as session:
            result = session.run("Qualify lead L-42")
            print(result.text)
            print(result.cache_hit)
    """

    def __init__(
        self,
        provider: str = "anthropic",
        *,
        system: str = "",
        immutable: str = "",
        capability: str = "",
        model: str | None = None,
        tools: list[Callable] | None = None,
        output_contract: OutputContract | None = None,
        session_backend: SessionBackend | None = None,
        api_key: str | None = None,
        **provider_kwargs: Any,
    ) -> None:
        self._provider_name = provider
        self._immutable = immutable or system
        self._capability = capability
        self._model = model or _DEFAULT_MODELS.get(provider, "claude-3-5-sonnet-20241022")
        self._output_contract = output_contract
        self._backend: SessionBackend = session_backend or MemoryBackend()

        # Tool registry (None if no tools)
        self._registry: ToolRegistry | None = (
            ToolRegistry(tools) if tools else None
        )

        # Provider adapter
        self._provider: ProviderBase = _make_provider(
            provider, self._model, api_key, **provider_kwargs
        )

        # Prompt builder
        self._prompt_builder = PromptBuilder(provider)

        # Capability enforcer (always present, even with empty capability text)
        self._capability_enforcer = CapabilityEnforcer(capability) if capability else None

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def session(self, **session_vars: Any) -> SessionContext:
        """Return a context manager that yields a ``Session``.

        ``session_vars`` are injected into the SESSION_STATE ICS layer on
        the first turn, providing initial context without requiring the
        caller to pre-format ICS text.

        Usage::

            with agent.session(lead_id="L-42") as session:
                result = session.run("Qualify this lead")
        """
        return SessionContext(self, session_vars)

    def run(self, task: str, **session_vars: Any) -> "RunResult":  # type: ignore[name-defined]
        """One-shot convenience: open a session, run one task, return the result.

        Equivalent to::

            with agent.session(**session_vars) as session:
                return session.run(task)
        """
        from ics_runtime.core.result import RunResult
        import uuid
        session = Session(self, str(uuid.uuid4()), session_vars)
        return session.run(task)

File: ics-runtime/java/src/main/java/io/ics/runtime/Agent.java
package io.ics.runtime;

import io.ics.runtime.backends.MemoryBackend;
import io.ics.runtime.backends.SessionBackend;
import io.ics.runtime.contracts.CapabilityEnforcer;
import io.ics.runtime.contracts.OutputContract;
import io.ics.runtime.prompt.PromptBuilder;
import io.ics.runtime.providers.AnthropicProvider;
import io.ics.runtime.providers.OpenAIProvider;
import io.ics.runtime.providers.ProviderBase;
import io.ics.runtime.tools.ToolDefinition;
import io.ics.runtime.tools.ToolRegistry;

import java.util.*;

/**
 * Top-level developer-facing object for ICS Runtime Java.
 *
 * <p>The Agent holds the static ICS configuration (immutable context, capability
 * declarations, tools, output contract) and spawns {@link Session} instances for
 * each conversation.  The Agent is stateless — all mutable state lives in the
 * Session.
 *
 * <p>Build via {@link Builder}:
 *
 * <pre>{@code
 * Agent agent = Agent.builder()
 *     .provider("anthropic")
 *     .immutable("You are a BFSI lead qualification assistant. ...")
 *     .capability("DENY: logging PII\nREQUIRE: risk category in every result")
 *     .model("claude-sonnet-4-6")
 *     .tool(crmLookupTool)
 *     .tool(eligibilityCheckTool)
 *     .outputContract(contract)
 *     .build();
 *
 * // One-shot run
 * RunResult r = agent.run("Qualify lead L-001");
 *
 * // Multi-turn session
 * try (Session session = agent.session(Map.of("lead_id", "L-001"))) {
 *     RunResult r1 = session.run("Qualify this lead");
 *     RunResult r2 = session.run("Explain the risk score");
 * }
 * }</pre>
 */
public final class Agent {

    private static final Map<String, String> DEFAULT_MODELS = Map.of(
            "anthropic", "claude-sonnet-4-6",
            "openai",    "gpt-4o"
    );

    private final String providerName;
    private final String immutable;
    private final String capability;
    private final String model;
    private final OutputContract outputContract;
    private final SessionBackend backend;
    private final ToolRegistry registry;
    private final ProviderBase provider;
    private final PromptBuilder promptBuilder;
    private final CapabilityEnforcer capabilityEnforcer;

    private Agent(Builder b) {
        this.providerName  = b.provider;
        this.immutable     = b.immutable != null ? b.immutable : b.system != null ? b.system : "";
        this.capability    = b.capability != null ? b.capability : "";
        this.model         = b.model != null ? b.model : DEFAULT_MODELS.getOrDefault(b.provider, "claude-sonnet-4-6");
        this.outputContract = b.outputContract;
        this.backend       = b.backend != null ? b.backend : new MemoryBackend();
        this.registry      = b.tools.isEmpty() ? null : new ToolRegistry(b.tools);
        this.provider      = buildProvider(b.provider, this.model, b.apiKey);
        this.promptBuilder = new PromptBuilder(b.provider);
        this.capabilityEnforcer = this.capability.isBlank() ? null : new CapabilityEnforcer(this.capability);
    }

    // ── Public API ───────────────────────────────────────────────────────────

    /**
     * Open a session and return it as an {@link AutoCloseable} (no-op close).
     *
     * @param sessionVars Variables injected into SESSION_STATE on the first turn
     */
    public Session session(Map<String, Object> sessionVars) {
        String sessionId = UUID.randomUUID().toString();
        return new Session(this, sessionId, sessionVars);
    }

    /** Convenience — open a session with no initial session vars. */
    public Session session() { return session(Map.of()); }

    /**
     * One-shot convenience — open a session, run one task, return the result.
     *
     * @param task        The user instruction
     * @param sessionVars Optional session context variables
     */
    public RunResult run(String task, Map<String, Object> sessionVars) {
        try (Session s = session(sessionVars)) {
            return s.run(task);
        }
    }

    /** Convenience — one-shot run with no session vars. */
    public RunResult run(String task) { return run(task, Map.of()); }

    // ── Package-visible accessors for Session ────────────────────────────────

    String getProviderName()               { return providerName; }
    String getImmutable()                  { return immutable; }
    String getCapability()                 { return capability; }
    String getModel()                      { return model; }
    OutputContract getOutputContract()     { return outputContract; }
    SessionBackend getBackend()            { return backend; }
    ToolRegistry getRegistry()             { return registry; }
    ProviderBase getProvider()             { return provider; }
    PromptBuilder getPromptBuilder()       { return promptBuilder; }
    CapabilityEnforcer getCapabilityEnforcer() { return capabilityEnforcer; }

    // ── Provider factory ─────────────────────────────────────────────────────

    private static ProviderBase buildProvider(String name, String model, String apiKey) {
        return switch (name) {
            case "anthropic" -> new AnthropicProvider(model, apiKey);
            case "openai"    -> new OpenAIProvider(model, apiKey);
            default -> throw new ICSRuntimeException(
                    "Unknown provider '" + name + "'. Supported: 'anthropic', 'openai'.");
        };
    }

    // ── Builder ──────────────────────────────────────────────────────────────

    public static Builder builder() { return new Builder(); }

    public static final class Builder {
        private String provider  = "anthropic";
        private String system    = null;
        private String immutable = null;
        private String capability = null;
        private String model     = null;
        private String apiKey    = null;
        private OutputContract outputContract = null;
        private SessionBackend backend = null;
        private final List<ToolDefinition> tools = new ArrayList<>();

        public Builder provider(String v)              { this.provider = v;   return this; }
        /** Alias for {@link #immutable(String)}. */
        public Builder system(String v)                { this.system = v;     return this; }
        public Builder immutable(String v)             { this.immutable = v;  return this; }
        public Builder capability(String v)            { this.capability = v; return this; }
        public Builder model(String v)                 { this.model = v;      return this; }
        public Builder apiKey(String v)                { this.apiKey = v;     return this; }
        public Builder outputContract(OutputContract v){ this.outputContract = v; return this; }
        public Builder backend(SessionBackend v)       { this.backend = v;    return this; }
        public Builder tool(ToolDefinition v)          { tools.add(v);        return this; }
        public Builder tools(List<ToolDefinition> v)   { tools.addAll(v);     return this; }
        public Agent build()                           { return new Agent(this); }
    }
}
